drivers/spi/spi_nxp_lpspi/spi_nxp_lpspi_dma.c

```diff
@@ -127,94 +188,124 @@ static inline int lpspi_dma_rxtx_load(const struct device *dev)
 	return dma_start(tx->dma_dev, tx->channel);
 }
 
-static int lpspi_dma_next_fill(const struct device *dev)
-{
-	struct lpspi_data *data = (struct lpspi_data *)dev->data;
-	struct spi_nxp_dma_data *dma_data = (struct spi_nxp_dma_data *)data->driver_data;
-	struct spi_dma_stream *rx = &dma_data->dma_rx;
-	struct spi_dma_stream *tx = &dma_data->dma_tx;
-
-	rx->chunk_done = false;
-	tx->chunk_done = false;
-
-	return lpspi_dma_rxtx_load(dev);
-}
-
 static void spi_mcux_dma_callback(const struct device *dev, void *arg, uint32_t channel, int status)
 {
+	/* arg directly holds the spi device */
 	const struct device *spi_dev = arg;
 	LPSPI_Type *base = (LPSPI_Type *)DEVICE_MMIO_NAMED_GET(spi_dev, reg_base);
 	struct lpspi_data *data = (struct lpspi_data *)spi_dev->data;
 	struct spi_nxp_dma_data *dma_data = (struct spi_nxp_dma_data *)data->driver_data;
 	struct spi_dma_stream *rx = &dma_data->dma_rx;
 	struct spi_dma_stream *tx = &dma_data->dma_tx;
 	struct spi_context *ctx = &data->ctx;
-	char debug_char;
+	char debug_char = (channel == dma_data->dma_tx.channel) ? 'T' : 'R';
+	int ret = 0;
 
 	if (status < 0) {
+		ret = status;
 		goto error;
-	} else {
-		/* don't care about positive values, normalize to "okay" = 0 */
-		status = 0;
 	}
 
-	if (channel == rx->channel) {
-		spi_context_update_rx(ctx, 1, rx->dma_blk_cfg.block_size);
-		debug_char = 'R';
-		rx->chunk_done = true;
-	} else if (channel == tx->channel) {
-		spi_context_update_tx(ctx, 1, tx->dma_blk_cfg.block_size);
-		debug_char = 'T';
-		tx->chunk_done = true;
-	} else {
-		/* invalid channel */
-		status = -EIO;
+	if (channel != dma_data->dma_tx.channel && channel != dma_data->dma_rx.channel) {
+		ret = -EIO;
 		goto error;
 	}
 
-	LOG_DBG("DMA %cX Block Complete", debug_char);
-
-	/* wait for the other channel to finish if needed */
-	if (!rx->chunk_done || !tx->chunk_done) {
-		return;
-	}
-
-
-	while ((IS_ENABLED(CONFIG_SOC_FAMILY_NXP_IMXRT) ||
-		IS_ENABLED(CONFIG_SOC_FAMILY_KINETIS)) &&
-		(base->SR & LPSPI_SR_MBF_MASK)) {
-		/* wait until module is idle */
-	}
-
-	if (spi_context_max_continuous_chunk(ctx) == 0) {
-		goto done;
-	}
-
-	status = lpspi_dma_next_fill(spi_dev);
-	if (status) {
-		goto error;
+	switch(dma_data->spi_transfer_state)
+	{
+		case LPSPI_TRANSFER_STATE_ONGOING:
```

Comment:
the point of splitting into states was to split the ongoing transfer into states, so having one state to describe the whole ongoing transfer makes this state paradigm not helpful

Reply:
so, what's your point? split into RX ongoing and TX ongoing?  this splitting will not reduce complexity of the code. you can try and give some example code.